edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironmentOperator.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes
+{
+    using System;
+    using System.Threading.Tasks;
+    using k8s;
+    using k8s.Models;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Rest;
+
+    public class KubernetesEnvironmentOperator : IKubernetesEnvironmentOperator
+    {
+        readonly IRuntimeInfoSource moduleStatusSource;
+        readonly IKubernetes client;
+        readonly string deviceNamespace;
+        Option<Watcher<V1Pod>> podWatch;
+
+        public KubernetesEnvironmentOperator(
+            string deviceNamespace,
+            IRuntimeInfoSource moduleStatusSource,
+            IKubernetes client)
+        {
+            this.deviceNamespace = deviceNamespace;
+            this.moduleStatusSource = moduleStatusSource;
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+            this.podWatch = Option.None<Watcher<V1Pod>>();
+        }
+
+        public void Start() => this.StartListPods();
+
+        public void Stop()
+        {
+            // TODO do we need lock here?
+            this.podWatch.ForEach(watch => watch.Dispose());
+        }
+
+        public void Dispose() => this.Stop();
+
+        void StartListPods() =>
+            this.client.ListNamespacedPodWithHttpMessagesAsync(this.deviceNamespace, watch: true)
+                .ContinueWith(this.OnListPodsCompleted);
+
+        async Task OnListPodsCompleted(Task<HttpOperationResponse<V1PodList>> task)
+        {
+            HttpOperationResponse<V1PodList> podListResp = await task;
+
+            this.podWatch = Option.Some(
+                podListResp.Watch<V1Pod>(
+                    onEvent: (type, item) =>
+                    {
+                        try
+                        {
+                            this.HandlePodChangedAsync(type, item);
+                        }
+                        catch (Exception ex) when (!ex.IsFatal())
+                        {
+                            Events.PodWatchFailed(ex);
+                        }
+                    },
+                    onClosed: () =>
+                    {
+                        Events.PodWatchClosed();
+
+                        // get rid of the current pod watch object since we got closed
+                        this.podWatch.ForEach(watch => watch.Dispose());
+                        this.podWatch = Option.None<Watcher<V1Pod>>();
+
+                        // kick off a new watch
+                        this.Start();
+                    },
+                    onError: Events.PodWatchFailed));
+        }
+
+        void HandlePodChangedAsync(WatchEventType type, V1Pod pod)
+        {
+            // if the pod doesn't have the module label set then we are not interested in it
+            if (!pod.Metadata.Labels.TryGetValue(Constants.K8sEdgeModuleLabel, out string podName))
```

Comment:
@darobs Should we continue to operate with pod name `edgeagent` or module name `edgeAgent` ?

Reply:
It looks like we can do either, as long as it's consistent. In reality, podname is module name.to_lower(), I thought to use module name because if there's a pod in the namespace that doesn't have the label, it's not a pod created for consumption of this watch. Once I got the module name, I continued to use it.